heft-plugins/heft-api-extractor-plugin/src/ApiExtractorPlugin.ts

```diff
@@ -0,0 +1,225 @@
+const TASK_CONFIG_RELATIVE_PATH: string = './config/api-extractor-task.json';
+const EXTRACTOR_CONFIG_FILENAME: typeof TApiExtractor.ExtractorConfig.FILENAME = 'api-extractor.json';
+const EXTRACTOR_CONFIG_RELATIVE_PATH: string = `./config/${EXTRACTOR_CONFIG_FILENAME}`;
+
+export interface IApiExtractorConfigurationResult {
+  apiExtractorPackage: typeof TApiExtractor;
+  apiExtractorConfiguration: TApiExtractor.ExtractorConfig;
+}
+
+export interface IApiExtractorTaskConfiguration {
+  /**
+   * If set to true, use the project's TypeScript compiler version for API Extractor's
+   * analysis. API Extractor's included TypeScript compiler can generally correctly
+   * analyze typings generated by older compilers, and referencing the project's compiler
+   * can cause issues. If issues are encountered with API Extractor's included compiler,
+   * set this option to true.
+   *
+   * This corresponds to API Extractor's `--typescript-compiler-folder` CLI option and
+   * `IExtractorInvokeOptions.typescriptCompilerFolder` API option. This option defaults to false.
+   */
+  useProjectTypescriptVersion?: boolean;
+}
+
+export default class ApiExtractorPlugin implements IHeftTaskPlugin {
+  private _apiExtractor: typeof TApiExtractor | undefined;
+  private _apiExtractorConfigurationFilePath: string | undefined | null = null;
+  private _apiExtractorTaskConfigurationFileLoader:
+    | ConfigurationFile<IApiExtractorTaskConfiguration>
+    | undefined;
+
+  public apply(taskSession: IHeftTaskSession, heftConfiguration: HeftConfiguration): void {
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Load up the configuration, but ignore if target files are missing, since we will be deleting
+      // them anyway.
+      const result: IApiExtractorConfigurationResult | undefined =
+        await this._getApiExtractorConfigurationAsync(
+          taskSession,
+          heftConfiguration,
+          /* ignoreMissingEntryPoint: */ true
+        );
+      if (result) {
+        this._includeOutputPathsInClean(cleanOptions, result.apiExtractorConfiguration);
+      }
+    });
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      const result: IApiExtractorConfigurationResult | undefined =
+        await this._getApiExtractorConfigurationAsync(taskSession, heftConfiguration);
+      if (result) {
+        await this._runApiExtractorAsync(
+          taskSession,
+          heftConfiguration,
+          runOptions,
+          result.apiExtractorPackage,
+          result.apiExtractorConfiguration
+        );
+      }
+    });
+  }
+
+  private async _getApiExtractorConfigurationFilePathAsync(
+    heftConfiguration: HeftConfiguration
+  ): Promise<string | undefined> {
+    // When null, we have yet to set the value. When undefined, the file could not be resolved.
```

Comment:
Stick this comment as a doc comment for the variable.

It may be cleaner to stick `null` in an `UNITIALIZED` constant. I think that exists in a few spots in Rush.

Reply:
Yeah, I like the constant. Means I don't need to add a doc comment :)